From 794c635aeaf6b2ff85296795ae81d4d5ad60ec3c Mon Sep 17 00:00:00 2001
From: joshjms <joshjms1607@gmail.com>
Date: Thu, 24 Apr 2025 09:08:51 +0000
Subject: [PATCH] Add defer-recover block to prevent panic when cc is nil

Cherry-pick commit fa38b54dd5db2f1f321ab6118cf55bec6124336b from pull
request #19792.

Signed-off-by: joshjms <joshjms1607@gmail.com>

[Backport to the address-based resolver in v3.5.21.]
(cherry picked from commit 55abfaafbee62a836d8ce178776dca202d5683b4)
Signed-off-by: cl2t <yihong.or@isrc.iscas.ac.cn>
---
 client/v3/internal/resolver/resolver.go | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/client/v3/internal/resolver/resolver.go b/client/v3/internal/resolver/resolver.go
index 3ee3cb8..d19c851 100644
--- a/client/v3/internal/resolver/resolver.go
+++ b/client/v3/internal/resolver/resolver.go
@@ -59,7 +59,7 @@ func (r *EtcdManualResolver) SetEndpoints(endpoints []string) {
 }
 
 func (r EtcdManualResolver) updateState() {
-	if r.CC != nil {
+	if getCC(r) != nil {
 		addresses := make([]resolver.Address, len(r.endpoints))
 		for i, ep := range r.endpoints {
 			addr, serverName := endpoint.Interpret(ep)
@@ -72,3 +72,13 @@ func (r EtcdManualResolver) updateState() {
 		r.UpdateState(state)
 	}
 }
+
+func getCC(r EtcdManualResolver) (cc resolver.ClientConn) {
+	defer func() {
+		if rec := recover(); rec != nil {
+			cc = nil
+		}
+	}()
+
+	return r.CC()
+}
